### DISCUSSION TITLE
Unable to rename table or column

### DISCUSSION THREAD

**Discussion by bduff9** [planetscale/discussion#122] [Bug Reports] [answered]
I was unable to find this on the docs or other questions.  If we have a change where we need to rename something (i.e. a table or column), PS applies it to the development branch without issue.  However, when attempting to deploy it to production, the actual change (the `ALTER TABLE` statement) is gone and instead, PS says it will drop the old and create the new.

Is this expected, i.e. there is no actual way to rename things without losing all the data?  Or is there a way around this or something?  While renaming is not common in our applications, it definitely does happen, and being able to do so without losing data is critical to our business.  Thank you!

**Comment by johnyerhot** [comment, marked as answer]
Hi @bduff9 👋  - good question.

The short answer is that PlanetScale does not support renaming tables or columns right now without data loss using `ALTER` statements. It's actually fairly complex on our end to be able to support fully in the way you're looking for and without application downtime.

In general performing a renaming operation will require downtime as code and the schema will be out of sync for a period of time. For some folks a maintenance window is fine, however for a production application it may not be acceptable.

That all said, what can be done is a process like this:

1. Create the new table
2. Deploy code that double writes to the new table and the old table
3. Backfill data from the old table to the new table
4. Switch reads to the new table once the data has been backfilled
5. Drop the old table

While that's a more complex process, it can be done such that it will result in no downtime. Part of the complexity on our end is we want to be able to effectively support table and column renames such that no downtime is needed. It's likely that we'll still suggest the above process but be able to abstract some bits of it away so that some of the double writes or reads are handled on our end.

However, we're still working it out 😄

**Reply by bduff9** [reply]
Thanks for your reply, @johnyerhot   I appreciate the detail.

Just to ensure I am understanding fully, the following is what is being suggested for this in cases where zero downtime is needed:
1. Add new tables and deploy in PlanetScale
2. Deploy new production code that writes to both new and old tables
3. Populate new tables with existing data from old tables
4. Deploy new production code that removes all references to old tables (both reads and writes)
5. Drop old tables and deploy in PlanetScale

While this requires extra work, it does seem relatively stable as none of the above steps are particularly time-sensitive, i.e. we can create the new tables at any point before deploying 2.  We can then deploy 2 at any point before doing 3, etc.

I am loving the idea and UI of PS so far so definitely eager to see how you improve things over time.  Thanks again for your response 👍

**Reply by johnyerhot** [reply]
You got it @bduff9 - it's a bit more complex on the surface, but is how a lot of big web applications have been handling zero downtime database changes and the process we aim to make as easy as possible to accomplish. That doesn't mean that we won't support renames via `ALTER` without data loss at some point but right now we do not.
